src/fal/planner/plan.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
+
+from typing import Iterator
+
+import networkx as nx
+
+
+def _find_subgraphs(graph: nx.DiGraph) -> Iterator[list[str]]:
+    # Initially topologically sort the graph, and split it
+    # into chunks by using the critical node approach (post-hooks
+    # are also a thing now).
+
+    # Then for each chunk, check whether they have the
+    # same set of ancestors. This is important because
+    # we want to group only the things that are in
+    # the same branch.
+
+    current_stack = []
+    allowed_ancestors = set()
+
+    def split() -> Iterator[list[str]]:
+        if len(current_stack) > 1:
+            yield current_stack.copy()
+        current_stack.clear()
+        allowed_ancestors.clear()
+
+    for node in nx.topological_sort(graph):
+        properties = graph.nodes[node]
+        if properties["kind"] == "python model":
```

Comment:
small question about this one; why are you assuming this to exist? dont we need a more structured node representation first? similar to what's discussed in the doc 

```
@dataclass
FalPythonModel(FalWorkUnit)
  dbt_model: ???
	script: FalScript
	hooks: List[FalHook]

@dataclass
DbtModel(FalWorkUnit)
	models: List[str]
	hooks: List[FalHook]

@dataclass
FalHook(FalWorkUnit)
	script: FalScript
```

Reply:
> small question about this one; why are you assuming this to exist? dont we need a more structured node representation first? similar to what's discussed in the doc

We convert to this representation when changing it from raw graph nodes into tasks at the scheduler. I think it is a bit easier to traverse/manipulate with the existing form on the planner. I think a better option would be ensuring these are issued at the graph construction stage, I'll also bundle the changes needed to the node graph to this PR. That should make this work, WDYT?